* * *

Statistics Introduction
-----------------------

Statistics is the science of analyzing, reviewing and conclude data.

Some basic statistical numbers include:

*   Mean, median and mode
*   Minimum and maximum value
*   Percentiles
*   Variance and Standard Devation
*   Covariance and Correlation
*   Probability distributions

The R language was developed by two statisticians. It has many built-in functionalities, in addition to libraries for the exact purpose of statistical analysis.

You will learn about these features, and how to use them in the next chapters.

**Tip:** If you are completely new to statistics, and want to learn more about it, visit our [Statistics Tutorial](https://www.w3schools.com/statistics/index.php).



# 2. R Data Set


* * *



Data Set
--------

A data set is a collection of data, often presented in a table.

There is a popular built-in data set in R called "**mtcars**" (Motor Trend Car Road Tests), which is retrieved from the 1974 Motor Trend US Magazine.

The `mtcars` dataset is a built-in dataset in R that contains measurements on 11 different attributes for 32 different cars. Here's what each of the columns in the dataset represents:

1. `mpg`: Miles/(US) gallon
2. `cyl`: Number of cylinders
3. `disp`: Displacement (cu.in.)
4. `hp`: Gross horsepower
5. `drat`: Rear axle ratio
6. `wt`: Weight (1000 lbs)
7. `qsec`: 1/4 mile time
8. `vs`: V/S
9. `am`: Transmission (0 = automatic, 1 = manual)
10. `gear`: Number of forward gears
11. `carb`: Number of carburetors

You can print the `mtcars` dataset in R by simply typing `mtcars` in the console. This will display the entire dataset in the console. 🚗📊

For more details on the `mtcars` dataset and other built-in datasets in R, you can explore additional resources such as [Statology](https://www.statology.org/mtcars-dataset-r/).

**Further info on mtcars:**

(1) A Complete Guide to the mtcars Dataset in R - Statology. https://www.statology.org/mtcars-dataset-r/.

(2) mtcars: Motor Trend Car Road Tests - R Package Documentation. https://rdrr.io/r/datasets/mtcars.html.

(3) R Statistics Data Set - W3Schools. https://www.w3schools.com/r/r_stat_data_set.asp.

(4) mtcars function - RDocumentation. https://www.rdocumentation.org/packages/datasets/versions/3.6.2/topics/mtcars.

In the examples below (and for the next chapters), we will use the `mtcars` data set, for statistical purposes:

### Example
```
mtcars
```
**Result:**
```

                     mpg cyl  disp  hp drat    wt  qsec vs am gear carb
Mazda RX4           21.0   6 160.0 110 3.90 2.620 16.46  0  1    4    4
Mazda RX4 Wag       21.0   6 160.0 110 3.90 2.875 17.02  0  1    4    4
Datsun 710          22.8   4 108.0  93 3.85 2.320 18.61  1  1    4    1
Hornet 4 Drive      21.4   6 258.0 110 3.08 3.215 19.44  1  0    3    1
Hornet Sportabout   18.7   8 360.0 175 3.15 3.440 17.02  0  0    3    2
Valiant             18.1   6 225.0 105 2.76 3.460 20.22  1  0    3    1
Duster 360          14.3   8 360.0 245 3.21 3.570 15.84  0  0    3    4
Merc 240D           24.4   4 146.7  62 3.69 3.190 20.00  1  0    4    2
Merc 230            22.8   4 140.8  95 3.92 3.150 22.90  1  0    4    2
Merc 280            19.2   6 167.6 123 3.92 3.440 18.30  1  0    4    4
Merc 280C           17.8   6 167.6 123 3.92 3.440 18.90  1  0    4    4
Merc 450SE          16.4   8 275.8 180 3.07 4.070 17.40  0  0    3    3
Merc 450SL          17.3   8 275.8 180 3.07 3.730 17.60  0  0    3    3
Merc 450SLC         15.2   8 275.8 180 3.07 3.780 18.00  0  0    3    3
Cadillac Fleetwood  10.4   8 472.0 205 2.93 5.250 17.98  0  0    3    4
Lincoln Continental 10.4   8 460.0 215 3.00 5.424 17.82  0  0    3    4
Chrysler Imperial   14.7   8 440.0 230 3.23 5.345 17.42  0  0    3    4
Fiat 128            32.4   4  78.7  66 4.08 2.200 19.47  1  1    4    1
Honda Civic         30.4   4  75.7  52 4.93 1.615 18.52  1  1    4    2
Toyota Corolla      33.9   4  71.1  65 4.22 1.835 19.90  1  1    4    1
Toyota Corona       21.5   4 120.1  97 3.70 2.465 20.01  1  0    3    1
Dodge Challenger    15.5   8 318.0 150 2.76 3.520 16.87  0  0    3    2
AMC Javelin         15.2   8 304.0 150 3.15 3.435 17.30  0  0    3    2
Camaro Z28          13.3   8 350.0 245 3.73 3.840 15.41  0  0    3    4
Pontiac Firebird    19.2   8 400.0 175 3.08 3.845 17.05  0  0    3    2
Fiat X1-9           27.3   4  79.0  66 4.08 1.935 18.90  1  1    4    1
Porsche 914-2       26.0   4 120.3  91 4.43 2.140 16.70  0  1    5    2
Lotus Europa        30.4   4  95.1 113 3.77 1.513 16.90  1  1    5    2
Ford Pantera L      15.8   8 351.0 264 4.22 3.170 14.50  0  1    5    4
Ferrari Dino        19.7   6 145.0 175 3.62 2.770 15.50  0  1    5    6
Maserati Bora       15.0   8 301.0 335 3.54 3.570 14.60  0  1    5    8
Volvo 142E          21.4   4 121.0 109 4.11 2.780 18.60  1  1    4    2
```
[Try it Yourself »](https://www.w3schools.com/r/tryr.asp?filename=demo_stat_ds_print)

* * *





In [ ]:
mtcars

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Mazda RX4           21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag       21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Datsun 710          22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive      21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout   18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2   
Valiant             18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1   
Duster 360          14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
Merc 240D           24.4 4   146.7  62 3.69 3.190 20.00 1  0  4    2   
Merc 230            22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2   
Merc 280            19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4   
Merc 280C           17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4   
Merc 450SE          16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3   
Merc 450SL          17.3 8   275.8 180 3.07 3.730 17.60 0  0  3    3   
Merc 450SLC         15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3   
Cadillac Fleetwood  10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4   
Lincoln Continental 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4   
Chrysler Imperial   14.7 8   440.0 230 3.23 5.345 17.42 0  0  3    4   
Fiat 128            32.4 4    78.7  66 4.08 2.200 19.47 1  1  4    1   
Honda Civic         30.4 4    75.7  52 4.93 1.615 18.52 1  1  4    2   
Toyota Corolla      33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1   
Toyota Corona       21.5 4   120.1  97 3.70 2.465 20.01 1  0  3    1   
Dodge Challenger    15.5 8   318.0 150 2.76 3.520 16.87 0  0  3    2   
AMC Javelin         15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2   
Camaro Z28          13.3 8   350.0 245 3.73 3.840 15.41 0  0  3    4   
Pontiac Firebird    19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2   
Fiat X1-9           27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1   
Porsche 914-2       26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2   
Lotus Europa        30.4 4    95.1 113 3.77 1.513 16.90 1  1  5    2   
Ford Pantera L      15.8 8   351.0 264 4.22 3.170 14.50 0  1  5    4   
Ferrari Dino        19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6   
Maserati Bora       15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8   
Volvo 142E          21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2

Information About the Data Set
------------------------------

You can use the question mark (`?`) to get information about the `mtcars` data set:

### Example

```  
?mtcars
```
Result:


<table summary="page for mtcars {datasets}"><tbody><tr><td>mtcars {datasets}</td><td>R Documentation</td></tr></tbody></table>

Motor Trend Car Road Tests
--------------------------

### Description

The data was extracted from the 1974 _Motor Trend_ US magazine, and comprises fuel consumption and 10 aspects of automobile design and performance for 32 automobiles (1973-74 models).

### Usage

mtcars

### Format

A data frame with 32 observations on 11 (numeric) variables.

<table summary="Rd table"><tbody><tr><td>[, 1]</td><td>mpg</td><td>Miles/(US) gallon</td></tr><tr><td>[, 2]</td><td>cyl</td><td>Number of cylinders</td></tr><tr><td>[, 3]</td><td>disp</td><td>Displacement (cu.in.)</td></tr><tr><td>[, 4]</td><td>hp</td><td>Gross horsepower</td></tr><tr><td>[, 5]</td><td>drat</td><td>Rear axle ratio</td></tr><tr><td>[, 6]</td><td>wt</td><td>Weight (1000 lbs)</td></tr><tr><td>[, 7]</td><td>qsec</td><td>1/4 mile time</td></tr><tr><td>[, 8]</td><td>vs</td><td>Engine (0 = V-shaped, 1 = straight)</td></tr><tr><td>[, 9]</td><td>am</td><td>Transmission (0 = automatic, 1 = manual)</td></tr><tr><td>[,10]</td><td>gear</td><td>Number of forward gears</td></tr><tr><td>[,11]</td><td>carb</td><td>Number of carburetors</td></tr></tbody></table>

### Note

Henderson and Velleman (1981) comment in a footnote to Table 1: 'Hocking \[original transcriber\]'s noncrucial coding of the Mazda's rotary engine as a straight six-cylinder engine and the Porsche's flat engine as a V engine, as well as the inclusion of the diesel Mercedes 240D, have been retained to enable direct comparisons to be made with previous analyses.'

### Source

Henderson and Velleman (1981), Building multiple regression models interactively. _Biometrics_, **37**, 391-411.

### Examples
```
require(graphics)
pairs(mtcars, main = "mtcars data", gap = 1/4)
coplot(mpg ~ disp | as.factor(cyl), data = mtcars,
       panel = panel.smooth, rows = 1)
## possibly more meaningful, e.g., for summary() or bivariate plots:
mtcars2 <- within(mtcars, {
   vs <- factor(vs, labels = c("V", "S"))
   am <- factor(am, labels = c("automatic", "manual"))
   cyl  <- ordered(cyl)
   gear <- ordered(gear)
   carb <- ordered(carb)
})
summary(mtcars2)
```
[Try it Yourself »](https://www.w3schools.com/r/tryr.asp?filename=demo_stat_ds_view)

* * *

* * *



In [ ]:
?mtcars

Get Information
---------------

Use the `dim()` function to find the dimensions of the data set, and the `names()` function to view the names of the variables:

### Example

Certainly! Let's break down the provided R code snippet:

```r
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use dim() to find the dimension (number of rows and columns) of the data set
dim(Data_Cars)

# Use names() to find the names of the variables (columns) in the data set
names(Data_Cars)
```
Result:
```
[1] 32 11
[1] "mpg"  "cyl"  "disp" "hp"   "drat" "wt"   "qsec" "vs"   "am"   "gear"
[1] "carb"
```

Here's what each part of the code does:

1. `Data_Cars <- mtcars`: This line creates a new variable called `Data_Cars` and assigns the entire `mtcars` data set to it. The `mtcars` data set contains information about various car models, including attributes like miles per gallon (mpg), horsepower, and weight.

2. `dim(Data_Cars)`: The `dim()` function returns a vector with two elements: the number of rows (observations) and the number of columns (variables) in the data set. In this case, it tells us how many cars (rows) and how many attributes (columns) are in the `Data_Cars` data frame.

3. `names(Data_Cars)`: The `names()` function retrieves the names of the columns (variables) in the data set. It provides a character vector containing the names of the attributes, such as "mpg," "cyl," "hp," etc.

By running this code, you'll get information about the dimensions of the data set (number of rows and columns) and the names of the variables within it. This is useful for understanding the structure of your data before performing any analyses or visualizations. 🚗📊


[Try it Yourself »](https://www.w3schools.com/r/tryr.asp?filename=demo_stat_ds_dim_names)



In [ ]:
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use dim() to find the dimension (number of rows and columns) of the data set
dim(Data_Cars)

# Use names() to find the names of the variables (columns) in the data set
names(Data_Cars)


[1] 32 11

[1] "mpg"  "cyl"  "disp" "hp"   "drat" "wt"   "qsec" "vs"   "am"   "gear"
[11] "carb"

Use the `rownames()` function to get the name of each row in the first column, which is the name of each car 🚗📊:

### Example

Certainly! Let's break down the provided R code snippet:

```r
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use rownames() to retrieve the row names of the data set
rownames(Data_Cars)
```

Here's what each part of the code does:

1. `Data_Cars <- mtcars`: This line creates a new variable called `Data_Cars` and assigns the entire `mtcars` data set to it. The `mtcars` data set contains information about various car models, including attributes like miles per gallon (mpg), horsepower, and weight.

2. `rownames(Data_Cars)`: The `rownames()` function retrieves the row names of the data set. It provides a character vector containing the names of the rows, which in this case are the names of the car models.

By running this code, you'll get a character vector containing the names of the car models in the `Data_Cars` data frame. Here's an example of what the output might look like:

```
 [1] "Mazda RX4"           "Mazda RX4 Wag"       "Datsun 710"         
 [4] "Hornet 4 Drive"      "Hornet Sportabout"   "Valiant"            
 [7] "Duster 360"          "Merc 240D"           "Merc 230"           
[10] "Merc 280"            "Merc 280C"           "Merc 450SE"         
[13] "Merc 450SL"          "Merc 450SLC"         "Cadillac Fleetwood"
[16] "Lincoln Continental" "Chrysler Imperial"   "Fiat 128"           
[19] "Honda Civic"         "Toyota Corolla"      "Toyota Corona"      
[22] "Dodge Challenger"    "AMC Javelin"         "Camaro Z28"         
[25] "Pontiac Firebird"    "Fiat X1-9"           "Porsche 914-2"      
[28] "Lotus Europa"        "Ford Pantera L"      "Ferrari Dino"       
[31] "Maserati Bora"       "Volvo 142E"  
```



Source:
(1) github.com. https://github.com/qaseleniumtesting01/GithubScannerRPackrat/tree/b0e4823f4868a7289715384c0c2d6a32df12fad8/README.md.

(2) github.com. https://github.com/geoffwi40/DataProducts/tree/8c68d217afa7e343f27e7b73536afb32930d6b8d/FuelEfficiency.md.

[Try it Yourself »](https://www.w3schools.com/r/tryr.asp?filename=demo_stat_ds_rownames)



In [ ]:
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use rownames() to retrieve the row names of the data set
rownames(Data_Cars)


[1] "Mazda RX4"           "Mazda RX4 Wag"       "Datsun 710"         
 [4] "Hornet 4 Drive"      "Hornet Sportabout"   "Valiant"            
 [7] "Duster 360"          "Merc 240D"           "Merc 230"           
[10] "Merc 280"            "Merc 280C"           "Merc 450SE"         
[13] "Merc 450SL"          "Merc 450SLC"         "Cadillac Fleetwood" 
[16] "Lincoln Continental" "Chrysler Imperial"   "Fiat 128"           
[19] "Honda Civic"         "Toyota Corolla"      "Toyota Corona"      
[22] "Dodge Challenger"    "AMC Javelin"         "Camaro Z28"         
[25] "Pontiac Firebird"    "Fiat X1-9"           "Porsche 914-2"      
[28] "Lotus Europa"        "Ford Pantera L"      "Ferrari Dino"       
[31] "Maserati Bora"       "Volvo 142E"

From the examples above, we have found out that the data set has **32** observations (Mazda RX4, Mazda RX4 Wag, Datsun 710, etc) and **11** variables (mpg, cyl, disp, etc).

A variable is defined as something that can be measured or counted.

Here is a brief explanation of the variables from the mtcars data set:

| Variable Name | Description |
| --- | --- |
| mpg | Miles/(US) Gallon |
| cyl | Number of cylinders |
| disp | Displacement |
| hp | Gross horsepower |
| drat | Rear axle ratio |
| wt | Weight (1000 lbs) |
| qsec | 1/4 mile time |
| vs | Engine (0 = V-shaped, 1 = straight) |
| am | Transmission (0 = automatic, 1 = manual) |
| gear | Number of forward gears |
| carb | Number of carburetors |

* * *



Print Variable Values
---------------------

If you want to print all values that belong to a variable, access the data frame by using the `$` sign, and the name of the variable (for example `cyl` (cylinders)):

### Example

```r
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Access the 'cyl' column from the Data_Cars data frame
Data_Cars$cyl
```

Here's what each part of the code does:

1. `Data_Cars <- mtcars`: This line creates a new variable called `Data_Cars` and assigns the entire `mtcars` data set to it. The `mtcars` data set contains information about various car models, including attributes like miles per gallon (mpg), horsepower, and weight.

2. `Data_Cars$cyl`: This expression accesses the column named "cyl" from the `Data_Cars` data frame. The `$` operator is used to extract a specific column by its name. In this case, it retrieves the values of the "cyl" column, which represents the number of cylinders in each car model.

   The result will be a numeric vector containing the cylinder values for all the cars in the data frame. For example:
   ```
   [1] 6 6 4 6 8 6 8 4 4 6 6 8 8 8 8 8 8 4 4 4 4 8 8 8 8 4 4 4 8 6 8 4
   ```

   Each value corresponds to the number of cylinders in a specific car model.

By running this code, you'll see the values of the "cyl" column from the `mtcars` data set. 🚗🔢

[Try it Yourself »](https://www.w3schools.com/r/tryr.asp?filename=demo_stat_ds_varval)

* * *



In [ ]:
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Access the 'cyl' column from the Data_Cars data frame
Data_Cars$cyl


[1] 6 6 4 6 8 6 8 4 4 6 6 8 8 8 8 8 8 4 4 4 4 8 8 8 8 4 4 4 8 6 8 4

Sort Variable Values
--------------------

To sort the values, use the `sort()` function:

### Example

```r
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use the sort() function to sort the 'cyl' column in ascending order
sort(Data_Cars$cyl)
```

Here's what each part of the code does:

1. `Data_Cars <- mtcars`: This line creates a new variable called `Data_Cars` and assigns the entire `mtcars` data set to it. The `mtcars` data set contains information about various car models, including attributes like miles per gallon (mpg), horsepower, and weight.

2. `sort(Data_Cars$cyl)`: This expression sorts the values in the "cyl" column of the `Data_Cars` data frame in ascending order. The `sort()` function is used to sort a vector or a column of a data frame. In this case, it sorts the values in the "cyl" column, which represents the number of cylinders in each car model.

   The result will be a numeric vector containing the cylinder values sorted in ascending order. For example:
   ```
   [1] 4 4 4 4 4 4 4 4 6 6 6 6 6 6 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8 8
   ```

   Each value corresponds to the number of cylinders in a specific car model, sorted in ascending order.

   From the examples above, we see that most cars have 4 and 8 cylinders.

By running this code, you'll see the values of the "cyl" column from the `mtcars` data set sorted in ascending order. 🚗🔢

[Try it Yourself »](https://www.w3schools.com/r/tryr.asp?filename=demo_stat_ds_sort)



* * *



In [ ]:
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use the sort() function to sort the 'cyl' column in ascending order
sort(Data_Cars$cyl)


[1] 4 4 4 4 4 4 4 4 4 4 4 6 6 6 6 6 6 6 8 8 8 8 8 8 8 8 8 8 8 8 8 8

Analyzing the Data
------------------

Now that we have some information about the data set, we can start to analyze it with some statistical numbers.

For example, we can use the `summary()` function to get a statistical summary of the data:

```r
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use the summary() function to summarize the data in the Data_Cars data frame
summary(Data_Cars)
```

Here's what each part of the code does:

1. `Data_Cars <- mtcars`: This line creates a new variable called `Data_Cars` and assigns the entire `mtcars` data set to it. The `mtcars` data set contains information about various car models, including attributes like miles per gallon (mpg), horsepower, and weight.

2. `summary(Data_Cars)`: This expression applies the `summary()` function to the `Data_Cars` data frame. The `summary()` function provides a concise summary of the data, including various descriptive statistics and information about each column (variable).

   The result will display summary statistics for each numeric column in the data frame. For example:
   ```
             mpg             cyl             disp             hp       
    Min.   :10.40   Min.   :4.000   Min.   : 71.1   Min.   : 52.0  
    1st Qu.:15.43   1st Qu.:4.000   1st Qu.:120.8   1st Qu.: 96.5  
    Median :19.20   Median :6.000   Median :196.3   Median :123.0  
    Mean   :20.09   Mean   :6.188   Mean   :230.7   Mean   :146.7  
    3rd Qu.:22.80   3rd Qu.:8.000   3rd Qu.:326.0   3rd Qu.:180.0  
    Max.   :33.90   Max.   :8.000   Max.   :472.0   Max.   :335.0  
       drat             wt             qsec             vs        
    Min.   :2.760   Min.   :1.513   Min.   :14.50   Min.   :0.0000  
    1st Qu.:3.080   1st Qu.:2.581   1st Qu.:16.89   1st Qu.:0.0000  
    Median :3.695   Median :3.325   Median :17.71   Median :0.0000  
    Mean   :3.597   Mean   :3.217   Mean   :17.85   Mean   :0.4375  
    3rd Qu.:3.920   3rd Qu.:3.610   3rd Qu.:18.90   3rd Qu.:1.0000  
    Max.   :4.930   Max.   :5.424   Max.   :22.90   Max.   :1.0000  
        am              gear            carb      
    Min.   :0.0000   Min.   :3.000   Min.   :1.000  
    1st Qu.:0.0000   1st Qu.:3.000   1st Qu.:2.000  
    Median :0.0000   Median :4.000   Median :2.000  
    Mean   :0.4062   Mean   :3.688   Mean   :2.812  
    3rd Qu.:1.0000   3rd Qu.:4.000   3rd Qu.:4.000  
    Max.   :1.0000   Max.   :5.000   Max.   :8.000  
   ```

   The summary includes minimum, maximum, mean, median, and quartile values for each numeric column. It also provides information about the data distribution.

By running this code, you'll get a comprehensive overview of the `mtcars` data set. 🚗📊.



In [ ]:
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use the summary() function to summarize the data in the Data_Cars data frame
summary(Data_Cars)


      mpg             cyl             disp             hp       
 Min.   :10.40   Min.   :4.000   Min.   : 71.1   Min.   : 52.0  
 1st Qu.:15.43   1st Qu.:4.000   1st Qu.:120.8   1st Qu.: 96.5  
 Median :19.20   Median :6.000   Median :196.3   Median :123.0  
 Mean   :20.09   Mean   :6.188   Mean   :230.7   Mean   :146.7  
 3rd Qu.:22.80   3rd Qu.:8.000   3rd Qu.:326.0   3rd Qu.:180.0  
 Max.   :33.90   Max.   :8.000   Max.   :472.0   Max.   :335.0  
      drat             wt             qsec             vs        
 Min.   :2.760   Min.   :1.513   Min.   :14.50   Min.   :0.0000  
 1st Qu.:3.080   1st Qu.:2.581   1st Qu.:16.89   1st Qu.:0.0000  
 Median :3.695   Median :3.325   Median :17.71   Median :0.0000  
 Mean   :3.597   Mean   :3.217   Mean   :17.85   Mean   :0.4375  
 3rd Qu.:3.920   3rd Qu.:3.610   3rd Qu.:18.90   3rd Qu.:1.0000  
 Max.   :4.930   Max.   :5.424   Max.   :22.90   Max.   :1.0000  
       am              gear            carb      
 Min.   :0.0000   Min.   :3.000  

Do not worry if you do not understand the output numbers. You will master them shortly.

The `summary()` function returns six statistical numbers for each variable:

*   Min
*   First quantile (percentile)
*   Median
*   Mean
*   Third quantile (percentile)
*   Max

We will cover all of them, along with other statistical numbers in the next chapters.

* * *

#Exercises for Data Sets

**Exercise 1: Fuel Efficiency Champion (Super Easy)**

**Instructions:**

1. **Explore the dataset:**
   - Print the first 5 rows of the `mtcars` dataset: `print(head(mtcars, 5))`
   - See the variable names: `names(mtcars)`
   - Get a statistical summary of the `mpg` (miles per gallon) variable: `summary(mtcars$mpg)`

2. **Find the most fuel-efficient car:**
   - Find the car with the highest `mpg` value: `which.max(mtcars$mpg)`
   - Use indexing to extract the model name and `mpg` value: `mtcars[which.max(mtcars$mpg), c("mpg", "model")]`

3. **Find the least fuel-efficient car:**
   - Find the car with the lowest `mpg` value: `which.min(mtcars$mpg)`
   - Extract the model name and `mpg` value: `mtcars[which.min(mtcars$mpg), c("mpg", "model")]`

**Challenge:** Create a bar chart showing the `mpg` for each car model. Highlight the most and least fuel-efficient cars in the chart.

**Exercise 2: Horsepower Powerhouse (Easy)**

**Instructions:**

1. **Filter cars with high horsepower:**
   - Create a subset of cars with horsepower (`hp`) above 200: `high_hp_cars <- mtcars[mtcars$hp > 200, ]`
   - Display the filtered subset: `print(high_hp_cars)`

2. **Calculate average horsepower:**
   - Calculate the average `hp` of all cars: `mean(mtcars$hp)`
   - Calculate the average `hp` of high-horsepower cars: `mean(high_hp_cars$hp)`

3. **Compare average horsepower:**
   - Print the average `hp` of all cars and high-horsepower cars side-by-side: `paste0("Average HP (all): ", mean(mtcars$hp), ", Average HP (high-HP): ", mean(high_hp_cars$hp))`

**Challenge:** Create a boxplot showing the distribution of `hp` for all cars and high-horsepower cars. Label the boxes accordingly.

**Exercise 3: Gear Up or Shift Down (Medium)**

**Instructions:**

1. **Analyze transmission types:**
   - Count the number of cars with automatic and manual transmissions: `table(mtcars$am)`
   - Calculate the percentage of cars with each transmission type: `prop.table(table(mtcars$am)) * 100`

2. **Relate gears to transmission:**
   - Calculate the average number of gears for automatic and manual cars: `tapply(mtcars$gear, mtcars$am, mean)`
   - Print the results: `print(paste0("Average gears (automatic): ", round(tapply(mtcars$gear, mtcars$am, mean)[1], 2), ", Average gears (manual): ", round(tapply(mtcars$gear, mtcars$am, mean)[2], 2)))`

3. **Explore fuel efficiency and gears:**
   - Calculate the average `mpg` for cars with different numbers of gears: `tapply(mtcars$mpg, mtcars$gear, mean)`
   - Create a scatter plot showing `mpg` on the y-axis and the number of gears on the x-axis. Use color to distinguish transmission types.

**Challenge:** Fit a linear regression model to predict `mpg` based on the number of gears and interpret the results.

#Solutions to R Data Set

## Solutions and Answers to R Programming Exercises with Explanations:

**Exercise 1: Fuel Efficiency Champion**

```r
# Explore the dataset
print(head(mtcars, 5))  # Print first 5 rows
names(mtcars)           # See variable names
summary(mtcars$mpg)     # Analyze mpg statistics

# Find most and least fuel-efficient cars
most_efficient <- mtcars[which.max(mtcars$mpg), c("mpg", "model")]
least_efficient <- mtcars[which.min(mtcars$mpg), c("mpg", "model")]

print(paste0("Most efficient: ", most_efficient["model"], " (", most_efficient["mpg"], " mpg)"))
print(paste0("Least efficient: ", least_efficient["model"], " (", least_efficient["mpg"], " mpg)"))

# Challenge: Bar chart with highlights
mpg_by_model <- data.frame(model = rownames(mtcars), mpg = mtcars$mpg)
library(ggplot2)
ggplot(mpg_by_model, aes(x = model, y = mpg)) +
  geom_bar(stat = "identity", fill = "lightblue") +
  labs(title = "Miles per Gallon by Car Model", x = "Car Model", y = "MPG") +
  geom_text(aes(x = model, y = mpg + 0.5, label = mpg), vjust = -0.5) +
  annotate("rect", xmin = most_efficient["model"], xmax = least_efficient["model"], ymin = min(mtcars$mpg), ymax = max(mtcars$mpg), alpha = 0.2, fill = "lightgray") +
  scale_x_discrete(limits = sort(mpg_by_model$model))  # Rotate labels for readability
```

**Exercise 2: Horsepower Powerhouse**

```r
# Filter cars with high horsepower
high_hp_cars <- mtcars[mtcars$hp > 200, ]
print(high_hp_cars)

# Calculate and compare average horsepower
avg_hp_all <- mean(mtcars$hp)
avg_hp_high <- mean(high_hp_cars$hp)

print(paste0("Average HP (all): ", round(avg_hp_all, 2), ", Average HP (high-HP): ", round(avg_hp_high, 2)))

# Challenge: Boxplot
ggplot(mtcars, aes(x = as.factor(am), y = hp)) +
  geom_boxplot() +
  labs(title = "Horsepower Distribution by Transmission", x = "Transmission (A: Automatic, M: Manual)", y = "Horsepower") +
  scale_x_discrete(labels = c("Automatic", "Manual"))
```

**Exercise 3: Gear Up or Shift Down**

```r
# Analyze transmission types
trans_counts <- table(mtcars$am)
trans_prop <- prop.table(trans_counts) * 100

print(paste0("Automatic: ", trans_counts[1], " (", round(trans_prop[1], 2), "%)"))
print(paste0("Manual: ", trans_counts[2], " (", round(trans_prop[2], 2), "%)"))

# Gears based on transmission
avg_gears <- tapply(mtcars$gear, mtcars$am, mean)

print(paste0("Average gears (automatic): ", round(avg_gears[1], 2), ", Average gears (manual): ", round(avg_gears[2], 2)))

# Explore fuel efficiency and gears
mpg_by_gear <- tapply(mtcars$mpg, mtcars$gear, mean)

ggplot(mtcars, aes(x = gear, y = mpg, color = as.factor(am))) +
  geom_point() +
  labs(title = "MPG by Number of Gears and Transmission", x = "Number of Gears", y = "MPG") +
  scale_color_discrete(labels = c("Automatic", "Manual"))
```

These solutions provide clear explanations and visualizations to enhance your understanding of the exercises. Feel free to experiment and modify them further to deepen your exploration of R and data analysis!

Remember to install the `ggplot2` library for the challenge plots.

Installing `ggplot2` on Google Colab is straightforward! Here's how you can do it:

**Method 1: Using the "Runtime" menu:**

1. Go to the **Runtime** menu in the top bar of your Colab notebook.
2. Select **Change runtime type**.
3. Choose a runtime type with `ggplot2` pre-installed, such as:
    - **Python 3 (TensorFlow 2.x)**
    - **R (3.6.3)**
    - **R (4.2.0)**
4. Click **Save and restart runtime**.

**Method 2: Using a magic command in your notebook:**

1. Within your notebook cell, type the following command and press Enter:

```
!pip install ggplot2
```

This will install `ggplot2` in the current runtime environment.

**Note:** If you're using Python, keep in mind that `ggplot2` is an R library. You'll need to use Python libraries like `rpy2` or `pynisher` to access R functionalities within your Python code.



#Max and Min

* * *

Max Min
-------

In the previous chapter, we introduced the **mtcars** data set. We will continue to use this data set throughout the next pages.

You learned from the [R Math](https://www.w3schools.com/r/r_math.asp) chapter that R has several built-in math functions. For example, the `min()` and `max()` functions can be used to find the lowest or highest value in a set:

### Example


```r
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Find the maximum value of the 'hp' (horsepower) column
max(Data_Cars$hp)

# Find the minimum value of the 'hp' (horsepower) column
min(Data_Cars$hp)
```

Here's what each part of the code does:

1. `Data_Cars <- mtcars`: This line creates a new variable called `Data_Cars` and assigns the entire `mtcars` data set to it. The `mtcars` data set contains information about various car models, including attributes like miles per gallon (mpg), horsepower, and weight.

2. `max(Data_Cars$hp)`: This expression calculates the maximum value of the "hp" column in the `Data_Cars` data frame. It finds the highest horsepower value among all the car models in the data set.

   The result will be a single numeric value representing the maximum horsepower. For example, if the highest horsepower in the data set is 335, the output will be:
   ```
   [1] 335
   ```

3. `min(Data_Cars$hp)`: Similarly, this expression calculates the minimum value of the "hp" column. It finds the lowest horsepower value among all the car models.

   The result will be a single numeric value representing the minimum horsepower. For example, if the lowest horsepower in the data set is 52, the output will be:
   ```
   [1] 52
   ```

By running this code, you'll see the maximum and minimum horsepower values in the `mtcars` data set. 🚗🔢


Result:

`[1] 335`  
`[1] 52`

[Try it Yourself »](https://www.w3schools.com/r/tryr.asp?filename=demo_stat_max_min)



In [ ]:
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Find the maximum value of the 'hp' (horsepower) column
max(Data_Cars$hp)

# Find the minimum value of the 'hp' (horsepower) column
min(Data_Cars$hp)


[1] 335

[1] 52

In [ ]:
mtcars

mpg  cyl disp  hp  drat wt    qsec  vs am gear carb
Mazda RX4           21.0 6   160.0 110 3.90 2.620 16.46 0  1  4    4   
Mazda RX4 Wag       21.0 6   160.0 110 3.90 2.875 17.02 0  1  4    4   
Datsun 710          22.8 4   108.0  93 3.85 2.320 18.61 1  1  4    1   
Hornet 4 Drive      21.4 6   258.0 110 3.08 3.215 19.44 1  0  3    1   
Hornet Sportabout   18.7 8   360.0 175 3.15 3.440 17.02 0  0  3    2   
Valiant             18.1 6   225.0 105 2.76 3.460 20.22 1  0  3    1   
Duster 360          14.3 8   360.0 245 3.21 3.570 15.84 0  0  3    4   
Merc 240D           24.4 4   146.7  62 3.69 3.190 20.00 1  0  4    2   
Merc 230            22.8 4   140.8  95 3.92 3.150 22.90 1  0  4    2   
Merc 280            19.2 6   167.6 123 3.92 3.440 18.30 1  0  4    4   
Merc 280C           17.8 6   167.6 123 3.92 3.440 18.90 1  0  4    4   
Merc 450SE          16.4 8   275.8 180 3.07 4.070 17.40 0  0  3    3   
Merc 450SL          17.3 8   275.8 180 3.07 3.730 17.60 0  0  3    3   
Merc 450SLC         15.2 8   275.8 180 3.07 3.780 18.00 0  0  3    3   
Cadillac Fleetwood  10.4 8   472.0 205 2.93 5.250 17.98 0  0  3    4   
Lincoln Continental 10.4 8   460.0 215 3.00 5.424 17.82 0  0  3    4   
Chrysler Imperial   14.7 8   440.0 230 3.23 5.345 17.42 0  0  3    4   
Fiat 128            32.4 4    78.7  66 4.08 2.200 19.47 1  1  4    1   
Honda Civic         30.4 4    75.7  52 4.93 1.615 18.52 1  1  4    2   
Toyota Corolla      33.9 4    71.1  65 4.22 1.835 19.90 1  1  4    1   
Toyota Corona       21.5 4   120.1  97 3.70 2.465 20.01 1  0  3    1   
Dodge Challenger    15.5 8   318.0 150 2.76 3.520 16.87 0  0  3    2   
AMC Javelin         15.2 8   304.0 150 3.15 3.435 17.30 0  0  3    2   
Camaro Z28          13.3 8   350.0 245 3.73 3.840 15.41 0  0  3    4   
Pontiac Firebird    19.2 8   400.0 175 3.08 3.845 17.05 0  0  3    2   
Fiat X1-9           27.3 4    79.0  66 4.08 1.935 18.90 1  1  4    1   
Porsche 914-2       26.0 4   120.3  91 4.43 2.140 16.70 0  1  5    2   
Lotus Europa        30.4 4    95.1 113 3.77 1.513 16.90 1  1  5    2   
Ford Pantera L      15.8 8   351.0 264 4.22 3.170 14.50 0  1  5    4   
Ferrari Dino        19.7 6   145.0 175 3.62 2.770 15.50 0  1  5    6   
Maserati Bora       15.0 8   301.0 335 3.54 3.570 14.60 0  1  5    8   
Volvo 142E          21.4 4   121.0 109 4.11 2.780 18.60 1  1  4    2

Now we know that the largest horsepower value in the set is **335**, and the lowest **52**.

We could take a look at the data set and try to find out which car these two values belongs to:

#### Observation of cars
```
                     mpg cyl  disp  hp drat    wt  qsec vs am gear carb
Mazda RX4           21.0   6 160.0 110 3.90 2.620 16.46  0  1    4    4
Mazda RX4 Wag       21.0   6 160.0 110 3.90 2.875 17.02  0  1    4    4
Datsun 710          22.8   4 108.0  93 3.85 2.320 18.61  1  1    4    1
Hornet 4 Drive      21.4   6 258.0 110 3.08 3.215 19.44  1  0    3    1
Hornet Sportabout   18.7   8 360.0 175 3.15 3.440 17.02  0  0    3    2
Valiant             18.1   6 225.0 105 2.76 3.460 20.22  1  0    3    1
Duster 360          14.3   8 360.0 245 3.21 3.570 15.84  0  0    3    4
Merc 240D           24.4   4 146.7  62 3.69 3.190 20.00  1  0    4    2
Merc 230            22.8   4 140.8  95 3.92 3.150 22.90  1  0    4    2
Merc 280            19.2   6 167.6 123 3.92 3.440 18.30  1  0    4    4
Merc 280C           17.8   6 167.6 123 3.92 3.440 18.90  1  0    4    4
Merc 450SE          16.4   8 275.8 180 3.07 4.070 17.40  0  0    3    3
Merc 450SL          17.3   8 275.8 180 3.07 3.730 17.60  0  0    3    3
Merc 450SLC         15.2   8 275.8 180 3.07 3.780 18.00  0  0    3    3
Cadillac Fleetwood  10.4   8 472.0 205 2.93 5.250 17.98  0  0    3    4
Lincoln Continental 10.4   8 460.0 215 3.00 5.424 17.82  0  0    3    4
Chrysler Imperial   14.7   8 440.0 230 3.23 5.345 17.42  0  0    3    4
Fiat 128            32.4   4  78.7  66 4.08 2.200 19.47  1  1    4    1
**Honda Civic         30.4   4  75.7  **52** 4.93 1.615 18.52  1  1    4    2
Toyota Corolla      33.9   4  71.1  65 4.22 1.835 19.90  1  1    4    1
Toyota Corona       21.5   4 120.1  97 3.70 2.465 20.01  1  0    3    1
Dodge Challenger    15.5   8 318.0 150 2.76 3.520 16.87  0  0    3    2
AMC Javelin         15.2   8 304.0 150 3.15 3.435 17.30  0  0    3    2
Camaro Z28          13.3   8 350.0 245 3.73 3.840 15.41  0  0    3    4
Pontiac Firebird    19.2   8 400.0 175 3.08 3.845 17.05  0  0    3    2
Fiat X1-9           27.3   4  79.0  66 4.08 1.935 18.90  1  1    4    1
Porsche 914-2       26.0   4 120.3  91 4.43 2.140 16.70  0  1    5    2
Lotus Europa        30.4   4  95.1 113 3.77 1.513 16.90  1  1    5    2
Ford Pantera L      15.8   8 351.0 264 4.22 3.170 14.50  0  1    5    4
Ferrari Dino        19.7   6 145.0 175 3.62 2.770 15.50  0  1    5    6
**Maserati Bora       15.0   8 301.0 **335** 3.54 3.570 14.60  0  1    5    8
Volvo 142E          21.4   4 121.0 109 4.11 2.780 18.60  1  1    4    2
```
By observing the table, it looks like the largest hp value belongs to a Maserati Bora, and the lowest belongs to a Honda Civic.

However, it is much easier (and safer) to let R find out this for us.

For example, we can use the `which.max()` and `which.min()` functions to find the index position of the max and min value in the table:

### Example



```r
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use the which.max() function to find the index position of the maximum value in the 'hp' column
which.max(Data_Cars$hp)

# Use the which.min() function to find the index position of the minimum value in the 'hp' column
which.min(Data_Cars$hp)
```
Result:
```
[1] 31   
[1] 19
```
Here's what each part of the code does:

1. `Data_Cars <- mtcars`: This line creates a new variable called `Data_Cars` and assigns the entire `mtcars` data set to it. The `mtcars` data set contains information about various car models, including attributes like miles per gallon (mpg), horsepower, and weight.

2. `which.max(Data_Cars$hp)`: This expression calculates the index position of the maximum value in the "hp" column of the `Data_Cars` data frame. It identifies the row corresponding to the car model with the highest horsepower.

   The result will be an integer value representing the position (row number) of the car model with the maximum horsepower. For example, if the highest horsepower is associated with the 31st car model, the output will be:
   ```
   [1] 31
   ```

3. `which.min(Data_Cars$hp)`: Similarly, this expression calculates the index position of the minimum value in the "hp" column. It identifies the row corresponding to the car model with the lowest horsepower.

   The result will be an integer value representing the position (row number) of the car model with the minimum horsepower. For example, if the lowest horsepower is associated with the 19th car model, the output will be:
   ```
   [1] 19
   ```

By running this code, you'll get the row numbers corresponding to the car models with the highest and lowest horsepower in the `mtcars` data set. 🚗🔢

[Try it Yourself »](https://www.w3schools.com/r/tryr.asp?filename=demo_stat_max_min2)




In [ ]:
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use the which.max() function to find the index position of the maximum value in the 'hp' column
which.max(Data_Cars$hp)

# Use the which.min() function to find the index position of the minimum value in the 'hp' column
which.min(Data_Cars$hp)


[1] 31

[1] 19

Or even better, combine `which.max()` and `which.min()` with the `rownames()` function to get the name of the car with the largest and smallest horsepower:

### Example

```r
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use the which.max() function to find the index position of the maximum value in the 'hp' column
# Use rownames() to retrieve the row name of the car with the highest horsepower
rownames(Data_Cars)[which.max(Data_Cars$hp)]

# Use the which.min() function to find the index position of the minimum value in the 'hp' column
# Use rownames() to retrieve the row name of the car with the lowest horsepower
rownames(Data_Cars)[which.min(Data_Cars$hp)]
```
Result:
```
[1] "Maserati Bora"   
[1] "Honda Civic"`
```

Here's what each part of the code does:

1. `Data_Cars <- mtcars`: This line creates a new variable called `Data_Cars` and assigns the entire `mtcars` data set to it. The `mtcars` data set contains information about various car models, including attributes like miles per gallon (mpg), horsepower, and weight.

2. `rownames(Data_Cars)[which.max(Data_Cars$hp)]`: This expression finds the car model with the highest horsepower by using the `which.max()` function to find the index position of the maximum value in the "hp" column of the `Data_Cars` data frame. It then uses the `rownames()` function to retrieve the row name of the car model at that index position.

   The result will be a character string representing the name of the car model with the highest horsepower. For example, if the car model with the highest horsepower is a Maserati Bora, the output will be:
   ```
   [1] "Maserati Bora"
   ```

3. `rownames(Data_Cars)[which.min(Data_Cars$hp)]`: Similarly, this expression finds the car model with the lowest horsepower by using the `which.min()` function to find the index position of the minimum value in the "hp" column of the `Data_Cars` data frame. It then uses the `rownames()` function to retrieve the row name of the car model at that index position.

   The result will be a character string representing the name of the car model with the lowest horsepower. For example, if the car model with the lowest horsepower is a Honda Civic, the output will be:
   ```
   [1] "Honda Civic"
   ```

By running this code, you'll get the names of the car models with the highest and lowest horsepower in the `mtcars` data set. 🚗🔢




[Try it Yourself »](https://www.w3schools.com/r/tryr.asp?filename=demo_stat_max_min3)




In [ ]:
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use the which.max() function to find the index position of the maximum value in the 'hp' column
# Use rownames() to retrieve the row name of the car with the highest horsepower
rownames(Data_Cars)[which.max(Data_Cars$hp)]

# Use the which.min() function to find the index position of the minimum value in the 'hp' column
# Use rownames() to retrieve the row name of the car with the lowest horsepower
rownames(Data_Cars)[which.min(Data_Cars$hp)]


[1] "Maserati Bora"

[1] "Honda Civic"

Now we know for sure:  
**Maserati Bora** is the car with the highest horsepower, and **Honda Civic** is the car with the lowest horsepower.

* * *

Outliers
--------

Max and min can also be used to detect **outliers**. An outlier is a data point that differs from rest of the observations.

Example of data points that could have been outliers in the **mtcars** data set:

*   If maximum of forward gears of a car was 11
*   If minimum of horsepower of a car was 0
*   If maximum weight of a car was 50 000 lbs

  

#Exercises on Max and Min

## Fun Exercises on Max & Min in R:

**Exercise 1: Movie Mania (Super Easy)**

Imagine you have a list of your favorite movies and their ratings (1-5). Use R to find the:

* **Highest-rated movie:** Which movie has the highest rating in your list?
* **Least-rated movie:** Which movie has the lowest rating in your list?

**Instructions:**

1. Create a list called `movies` with each element containing two values: movie title and rating (e.g., `c("Spider-Man", 5)`).
2. Use `max()` and `min()` functions with the `movies` list to find the highest and lowest ratings.
3. Modify the code to also print the movie titles along with their ratings.

**Challenge:** Use string manipulation (e.g., `substr()`) to extract only the movie titles and display them in a more user-friendly format.

**Learner Code Snippet:**

```r
# Create your movie list (replace with your own!)
movies <- c("Spider-Man", 5, "The Godfather", 4, "Spirited Away", 5)

# Find highest and lowest ratings
highest_rating <- max(movies[2:length(movies)])
lowest_rating <- min(movies[2:length(movies)])

# Print results (modify to display titles)
print(paste0("Highest rated: ", highest_rating))
print(paste0("Lowest rated: ", lowest_rating))
```

**Exercise 2: Sleep Tracker (Easy)**

You recorded your sleep quality score (1-10) for each day of the week (Monday-Sunday). Use R to find:

* **Best sleep day:** Which day had the highest average sleep quality?
* **Worst sleep day:** Which day had the lowest average sleep quality?

**Instructions:**

1. Create a vector called `sleep_scores` with your daily sleep scores.
2. Create a character vector called `days` with the corresponding days of the week.
3. Use `tapply()` to calculate the average sleep score for each day.
4. Use `which.max()` and `which.min()` with the result from `tapply()` to find the best and worst sleep days.

**Challenge:** Plot a bar chart showing the average sleep score for each day. Highlight the best and worst days visually.

**Learner Code Snippet:**

```r
# Enter your sleep scores for each day (1-10)
sleep_scores <- c(7, 6, 8, 8, 7, 9, 9)

# Days of the week
days <- c("Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday")

# Calculate average sleep score per day
avg_sleep_per_day <- tapply(sleep_scores, days, mean)

# Find best and worst sleep days
best_day <- days[which.max(avg_sleep_per_day)]
worst_day <- days[which.min(avg_sleep_per_day)]

# Print results
print(paste0("Best sleep day: ", best_day))
print(paste0("Worst sleep day: ", worst_day))
```


**Exercise 3: Video Game Champion (Medium)**

You play various video games and track your playtime in hours. Use R to determine:

* **Most played genre:** Which genre of video games did you play the most in total?
* **Least played genre:** Which genre did you play the least in total?

**Instructions:**

1. Create a list called `games` where each element represents a game with its genre and playtime (e.g., `list(genre="Action", playtime=20)`).
2. Use `tapply()` to sum the playtime for each genre in the `games` list.
3. Use `which.max()` and `which.min()` with the result from `tapply()` to find the most and least played genres.

**Challenge:** Create a stacked bar chart showing the total playtime for each genre. Order the bars based on playtime (highest to lowest).

**Learner Code Snippet:**

```r
# Create your game list (replace with your own!)
games <- list(
  genre="Action", playtime=20,
  genre="RPG", playtime=15,
  genre="Puzzle", playtime=10
)

# Calculate total playtime per genre
total_playtime_per_genre <- tapply(unlist(sapply(games, "[[2]]")), unlist(sapply(games, "[[1]]")), sum)

# Find most and least played genres
most_played <- names(total_playtime_per_genre)[which.max(total_playtime_per_genre)]
least_played <- names(total_playtime_per_genre)[which.min(total_playtime_per_genre)]

# Print results
print(paste0("Most played genre: ", most_played))
print(paste0("Least played genre: ", least_played))

# Challenge: Stacked bar chart with sorting (modify colors and labels)
genres <- names(total_playtime_per_genre)
playtime <- sort(total_playtime_per_genre, decreasing = TRUE)
barplot(playtime, col = c("red", "orange", "green"), names.arg = genres, main = "Total Playtime by Genre", xlab = "Genre", ylab = "Hours Played")
```

Remember to replace the sample game list with your own data and personalize the challenge steps according to your preferences!

# Solutions to Exercises for Max and Min

## Fun Exercises on Max & Min in R (Solutions with Explanations):

**Exercise 1: Movie Mania (Easy)**

**Solution:**

```r
# Create your movie list (replace with your own!)
movies <- c("Spider-Man", 5, "The Godfather", 4, "Spirited Away", 5)

# Find highest and lowest ratings
highest_rating <- max(movies[2:length(movies)])
lowest_rating <- min(movies[2:length(movies)])

# Extract movie titles (modify for different list structures)
movie_titles <- substr(movies, 1, nchar(movies) - 1)

# Print results with titles and ratings
print(paste0("Highest rated: ", movie_titles[which.max(movies[2:length(movies)])], " (", highest_rating, ")"))
print(paste0("Lowest rated: ", movie_titles[which.min(movies[2:length(movies)])], " (", lowest_rating, ")"))
```

**Challenge:** This solution extracts movie titles using `substr()` by assuming titles start at index 1 and ratings are at the end. Modify it for lists with different structures.

**Exercise 2: Sleep Tracker (Easy)**

**Solution:**

```r
# Enter your sleep scores for each day (1-10)
sleep_scores <- c(7, 6, 8, 8, 7, 9, 9)

# Days of the week
days <- c("Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday")

# Calculate average sleep score per day
avg_sleep_per_day <- tapply(sleep_scores, days, mean)

# Find best and worst sleep days
best_day <- days[which.max(avg_sleep_per_day)]
worst_day <- days[which.min(avg_sleep_per_day)]

# Print results
print(paste0("Best sleep day: ", best_day, " (average score: ", round(avg_sleep_per_day[which.max(avg_sleep_per_day)], 2), ")"))
print(paste0("Worst sleep day: ", worst_day, " (average score: ", round(avg_sleep_per_day[which.min(avg_sleep_per_day)], 2), ")"))

# Challenge: Stacked bar chart (modify colors and labels)
barplot(avg_sleep_per_day, col = c("lightblue", "lightgray"), names.arg = days, main = "Average Sleep Quality by Day", xlab = "Day", ylab = "Average Score")
text(avg_sleep_per_day + 0.1, avg_sleep_per_day, pos = 4, labels = round(avg_sleep_per_day, 1))  # Add data labels
abline(h = mean(avg_sleep_per_day), col = "red", lty = 2)  # Add average sleep line
```

**Challenge:** This solution adds data labels and highlights the average sleep quality with a horizontal line. Customize colors and labels as desired.

**Exercise 3: Video Game Champion (Medium)**

**Solution:**

```r
# Create your game list (replace with your own!)
games <- list(
  genre="Action", playtime=20,
  genre="RPG", playtime=15,
  genre="Puzzle", playtime=10
)

# Calculate total playtime per genre
total_playtime_per_genre <- tapply(unlist(sapply(games, "[[2]]")), unlist(sapply(games, "[[1]]")), sum)

# Find most and least played genres
most_played <- names(total_playtime_per_genre)[which.max(total_playtime_per_genre)]
least_played <- names(total_playtime_per_genre)[which.min(total_playtime_per_genre)]

# Print results
print(paste0("Most played genre: ", most_played, " (", total_playtime_per_genre[which.max(total_playtime_per_genre)], " hours)"))
print(paste0("Least played genre: ", least_played, " (", total_playtime_per_genre[which.min(total_playtime_per_genre)], " hours)"))

# Challenge: Stacked bar chart with sorting (modify colors and labels)
genres <- names(total_playtime_per_genre)
playtime <- sort(total_playtime_per_genre, decreasing
```

# Mean, Mode, Median

* * *

Mean, Median, and Mode
----------------------

In statistics, there are often three values that interests us:

*   **Mean** - The average value
*   [**Median**](https://www.w3schools.com/r/r_stat_median.asp) - The middle value
*   [**Mode**](https://www.w3schools.com/r/r_stat_mode.asp) - The most common value

* * *

Mean
----

To calculate the average value (mean) of a variable from the `mtcars` data set, find the sum of all values, and divide the sum by the number of values.

#### Sorted observation of wt (weight)

<table><tbody><tr><td>1.513</td><td>1.615</td><td>1.835</td><td>1.935</td><td>2.140</td><td>2.200</td><td>2.320</td><td>2.465</td></tr><tr><td>2.620</td><td>2.770</td><td>2.780</td><td>2.875</td><td>3.150</td><td>3.170</td><td>3.190</td><td>3.215</td></tr><tr><td>3.435</td><td>3.440</td><td>3.440</td><td>3.440</td><td>3.460</td><td>3.520</td><td>3.570</td><td>3.570</td></tr><tr><td>3.730</td><td>3.780</td><td>3.840</td><td>3.845</td><td>4.070</td><td>5.250</td><td>5.345</td><td>5.424</td></tr></tbody></table>

Luckily for us, the `mean()` function in R can do it for you:

### Example

Find the average weight (`wt`) of a car:


```r
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Calculate the mean (average) value of the 'wt' (weight) column
mean(Data_Cars$wt)
```

Result:

`[1] 3.21725`

Here's what each part of the code does:

1. `Data_Cars <- mtcars`: This line creates a new variable called `Data_Cars` and assigns the entire `mtcars` data set to it. The `mtcars` data set contains information about various car models, including attributes like miles per gallon (mpg), horsepower, and weight.

2. `mean(Data_Cars$wt)`: This expression calculates the **mean (average) value** of the "wt" (weight) column in the `Data_Cars` data frame. It computes the sum of all weight values and divides it by the total number of observations (car models).

   The result will be a single numeric value representing the average weight of all the cars in the data set. For example, if the average weight is 3.21725 (units depend on the original data), the output will be:
   ```
   [1] 3.21725
   ```

By running this code, you'll get the average weight of the cars in the `mtcars` data set. 🚗🔢




[Try it Yourself »](https://www.w3schools.com/r/tryr.asp?filename=demo_stat_mean)

* * *


In [ ]:
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Calculate the mean (average) value of the 'wt' (weight) column
mean(Data_Cars$wt)


[1] 3.21725

* * *

Median
------

The median value is the value in the middle, after you have sorted all the values.

If we take a look at the values of the `wt` variable (from the `mtcars` data set), we will see that there are two numbers in the middle:

#### Sorted observation of wt (weight)

<table><tbody><tr><td>1.513</td><td>1.615</td><td>1.835</td><td>1.935</td><td>2.140</td><td>2.200</td><td>2.320</td><td>2.465</td></tr><tr><td>2.620</td><td>2.770</td><td>2.780</td><td>2.875</td><td>3.150</td><td>3.170</td><td>3.190</td><td>3.215</td></tr><tr><td>3.435</td><td>3.440</td><td>3.440</td><td>3.440</td><td>3.460</td><td>3.520</td><td>3.570</td><td>3.570</td></tr><tr><td>3.730</td><td>3.780</td><td>3.840</td><td>3.845</td><td>4.070</td><td>5.250</td><td>5.345</td><td>5.424</td></tr></tbody></table>

**Note:** If there are two numbers in the middle, you must divide the sum of those numbers by two, to find the median.

Luckily, R has a function that does all of that for you: Just use the `median()` function to find the middle value:

### Example

Find the mid point value of weight (`wt`):

```r
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Calculate the median value of the 'wt' (weight) column
median(Data_Cars$wt)
```

Result:

`[1] 3.325`

Here's what each part of the code does:

1. `Data_Cars <- mtcars`: This line creates a new variable called `Data_Cars` and assigns the entire `mtcars` data set to it. The `mtcars` data set contains information about various car models, including attributes like miles per gallon (mpg), horsepower, and weight.

2. `median(Data_Cars$wt)`: This expression calculates the **median value** of the "wt" (weight) column in the `Data_Cars` data frame. The median is the middle value when all the weight values are sorted in ascending order.

   The result will be a single numeric value representing the median weight of all the cars in the data set. For example, if the median weight is 3.325 (units depend on the original data), the output will be:
   ```
   [1] 3.325
   ```

By running this code, you'll get the median weight of the cars in the `mtcars` data set. 🚗🔢





[Try it Yourself »](https://www.w3schools.com/r/tryr.asp?filename=demo_stat_median)

* * *



In [ ]:
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Calculate the median value of the 'wt' (weight) column
median(Data_Cars$wt)


[1] 3.325

* * *

Mode
----

The mode value is the value that appears the most number of times.

R does not have a function to calculate the mode. However, we can create our own function to find it.

If we take a look at the values of the `wt` variable (from the `mtcars` data set), we will see that the numbers **3.440** are often shown:

#### Sorted observation of wt (weight)

<table><tbody><tr><td>1.513</td><td>1.615</td><td>1.835</td><td>1.935</td><td>2.140</td><td>2.200</td><td>2.320</td><td>2.465</td></tr><tr><td>2.620</td><td>2.770</td><td>2.780</td><td>2.875</td><td>3.150</td><td>3.170</td><td>3.190</td><td>3.215</td></tr><tr><td>3.435</td><td>3.440</td><td>3.440</td><td>3.440</td><td>3.460</td><td>3.520</td><td>3.570</td><td>3.570</td></tr><tr><td>3.730</td><td>3.780</td><td>3.840</td><td>3.845</td><td>4.070</td><td>5.250</td><td>5.345</td><td>5.424</td></tr></tbody></table>

Instead of counting it ourselves, we can use the following code to find the mode:

### Example


```r
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use the table() function to count the frequency of each weight value
# Use the sort() function to sort the frequency table in descending order
# Use the names() function to retrieve the name of the car model with the highest weight
names(sort(-table(Data_Cars$wt)))[1]
```
Result:

`[1] "3.44"`

Here's what each part of the code does:

1. `Data_Cars <- mtcars`: This line creates a new variable called `Data_Cars` and assigns the entire `mtcars` data set to it. The `mtcars` data set contains information about various car models, including attributes like miles per gallon (mpg), horsepower, and weight.

2. `table(Data_Cars$wt)`: This expression uses the `table()` function to count the frequency of each weight value in the "wt" column of the `Data_Cars` data frame. It creates a frequency table that shows how many times each weight value appears in the data set.

3. `sort(-table(Data_Cars$wt))`: This expression sorts the frequency table in descending order by using the `sort()` function. The `-` sign is used to sort the table in reverse order (i.e., from highest to lowest frequency).

4. `names(sort(-table(Data_Cars$wt)))[1]`: Finally, this expression retrieves the name of the car model with the highest weight by using the `names()` function to extract the first element of the sorted frequency table.

   The result will be a character string representing the name of the car model with the highest weight. For example, if the car model with the highest weight is a Lincoln Continental, the output will be:
   ```
   [1] "Lincoln Continental"
   ```

By running this code, you'll get the name of the car model with the highest weight in the `mtcars` data set. 🚗🔢




[Try it Yourself »](https://www.w3schools.com/r/tryr.asp?filename=demo_stat_mode)

From the example above, we now know that the number that appears the most number of times in mtcars wt variable is **3.44** or **3.440 lbs**.

* * *



In [ ]:
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use the table() function to count the frequency of each weight value
# Use the sort() function to sort the frequency table in descending order
# Use the names() function to retrieve the name of the car model with the highest weight
names(sort(-table(Data_Cars$wt)))[1]


[1] "3.44"

In [ ]:
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Use the table() function to count the frequency of each weight value
# Use the sort() function to sort the frequency table in descending order
# Use the names() function to retrieve the name of the car model with the highest weight
sort(-table(Data_Cars$wt))[1]

3.44 
  -3

In [ ]:
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars
# to display the name of the row with  the highest frequency item
car_names <- row.names(Data_Cars[which.max(Data_Cars$wt), ])
car_names

[1] "Lincoln Continental"

In [ ]:
library(DescTools)
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars$wt

Mode(Data_Cars)

[1] 3.44
attr(,"freq")
[1] 3

In [ ]:
install.packages("DescTools")
library(DescTools)

# Sample data
my_data <- c(3, 5, 2, 5, 1, 3, 2, 5, 4, 3)

# Calculate mode
mode <- Mode(my_data)

# Print the mode
print(mode)


Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘proxy’, ‘rootSolve’, ‘e1071’, ‘lmom’, ‘mvtnorm’, ‘expm’, ‘Rcpp’, ‘Exact’, ‘gld’




[1] 3 5
attr(,"freq")
[1] 3


# Exercises for Mean, Mode, Median

**Exercise 1: Speedy Cars - Super Easy**

**Instructions:**

1. **Load the `mtcars` dataset:**
   ```R
   # Load the mtcars dataset
   library(datasets)
   data(mtcars)
   ```

2. **Calculate the average speed (mpg):**
   ```R
   # Calculate the mean mpg
   average_mpg <- mean(mtcars$mpg)
   ```

3. **Find the fastest car:**
   ```R
   # Find the car with the highest mpg
   fastest_car <- mtcars[which.max(mtcars$mpg), ]
   ```

4. **Find the slowest car:**
   ```R
   # Find the car with the lowest mpg
   slowest_car <- mtcars[which.min(mtcars$mpg), ]
   ```

5. **Challenge:** Create a bar chart showing the top 5 fastest and slowest cars with their mpg values. Color-code the bars to distinguish the "speed demons" from the "fuel sippers."

**Exercise 2: Horsepower Heroes - Easy**

**Instructions:**

1. **Identify powerful cars:**
   ```R
   # Filter cars with horsepower (hp) above 200
   powerful_cars <- mtcars[mtcars$hp > 200, ]
   ```

2. **Explore average horsepower:**
   ```R
   # Calculate average hp for all cars
   average_hp_all <- mean(mtcars$hp)
   # Calculate average hp for powerful cars
   average_hp_powerful <- mean(powerful_cars$hp)
   ```

3. **Compare horsepower distributions:**
   ```R
   # Challenge: Create a violin plot showing the hp distribution for all cars and powerful cars. Label the distributions accordingly.
   ```

**Exercise 3: Gear Up or Shift Down - Medium**

**Instructions:**

1. **Analyze transmission types:**
   ```R
   # Count automatic and manual cars
   transmission_counts <- table(mtcars$am)
   ```

2. **Relate gears to transmission:**
   ```R
   # Calculate average gears for each transmission type
   average_gears <- tapply(mtcars$gear, mtcars$am, mean)
   ```

3. **Explore mpg and gears:**
   ```R
   # Calculate average mpg for different gear counts
   mpg_by_gear <- tapply(mtcars$mpg, mtcars$gear, mean)
   ```

4. **Challenge:** Create a scatter plot showing the relationship between mpg and the number of gears for each transmission type. Use color to distinguish the transmission types.

Remember to replace the empty lines with the corresponding code to complete the exercises. Use clear comments and variable names to explain your steps. Have fun exploring the `mtcars` dataset and learning R along the way!

#Solutions to Mean, Mode, Median

Absolutely! Here are the solutions and answers for the exercises and challenges, incorporating comments and explanations:

**Exercise 1: Speedy Cars**

```r
# Load the mtcars dataset
library(datasets)
data(mtcars)

# Calculate the average speed (mpg)
average_mpg <- mean(mtcars$mpg)
print(paste0("Average MPG:", round(average_mpg, 2)))

# Find the fastest car
fastest_car <- mtcars[which.max(mtcars$mpg), ]
print(paste0("Fastest car:", fastest_car$model, "(", fastest_car$mpg, " mpg)"))

# Find the slowest car
slowest_car <- mtcars[which.min(mtcars$mpg), ]
print(paste0("Slowest car:", slowest_car$model, "(", slowest_car$mpg, " mpg)"))

# Challenge: Bar chart with colors
top_5_fastest <- mtcars[order(mtcars$mpg, decreasing = TRUE)[1:5], ]
top_5_slowest <- mtcars[order(mtcars$mpg)[1:5], ]

ggplot(rbind(top_5_fastest, top_5_slowest), aes(x = model, y = mpg, fill = factor(mpg > average_mpg))) +
  geom_bar(stat = "identity") +
  labs(title = "Top 5 Fastest and Slowest Cars (MPG)", x = "Car Model", y = "MPG") +
  scale_fill_discrete(labels = c("Slowest", "Fastest")) +
  theme_bw()  # Black and white theme for clarity
```

**Exercise 2: Horsepower Heroes**

```r
# Filter cars with horsepower (hp) above 200
powerful_cars <- mtcars[mtcars$hp > 200, ]

# Calculate average hp for all cars
average_hp_all <- mean(mtcars$hp)
# Calculate average hp for powerful cars
average_hp_powerful <- mean(powerful_cars$hp)

print(paste0("Average HP (all):", round(average_hp_all, 2), ", Average HP (powerful):", round(average_hp_powerful, 2)))

# Challenge: Violin plot
library(ggplot2)
ggplot(mtcars, aes(x = as.factor(am), y = hp)) +
  geom_violin() +
  labs(title = "Horsepower Distribution by Transmission", x = "Transmission (A: Automatic, M: Manual)", y = "Horsepower") +
  scale_x_discrete(labels = c("Automatic", "Manual"))
```

**Exercise 3: Gear Up or Shift Down**

```r
# Analyze transmission types
transmission_counts <- table(mtcars$am)

# Calculate average gears for each transmission type
average_gears <- tapply(mtcars$gear, mtcars$am, mean)

print(paste0("Automatic:", transmission_counts[1], " cars, Avg gears:", round(average_gears[1], 2)))
print(paste0("Manual:", transmission_counts[2], " cars, Avg gears:", round(average_gears[2], 2)))

# Explore mpg and gears
mpg_by_gear <- tapply(mtcars$mpg, mtcars$gear, mean)

# Challenge: Scatter plot with colors
ggplot(mtcars, aes(x = gear, y = mpg, color = as.factor(am))) +
  geom_point() +
  labs(title = "MPG by Number of Gears and Transmission", x = "Number of Gears", y = "MPG") +
  scale_color_discrete(labels = c("Automatic", "Manual"))
```

Remember to install the `ggplot2` library for the challenge plots. These solutions showcase clear explanations and informative visuals to deepen your understanding of the concepts. Feel free to experiment and modify them further to explore the fascinating world of data analysis with R!

#R Quantiles

### Main Concepts

Let's clear up the main concepts first by breaking down the differences between **quantiles**, **percentiles**, **percentages**, and **quartiles**:

1. **Quantiles**:
   - Quantiles divide a dataset into equal parts.
   - They can range from any value to any other value.
   - Examples of quantiles include quartiles (4-quantiles), quintiles (5-quantiles), deciles (10-quantiles), and percentiles (100-quantiles).
   - Quantiles are typically expressed as decimal values and range from 0 to 1.

2. **Percentiles**:
   - Percentiles are a specific type of quantile.
   - They divide a dataset into 100 equal parts.
   - Percentiles range from 0 to 100.
   - For example, the 90th percentile represents the value below which 90% of the data falls.

3. **Percentages**:
   - Percentages are not directly related to quantiles.
   - They express a proportion or relative part of a whole.
   - Percentages are always expressed as a percentage (e.g., 50%, 75%, etc.).
   - For example, if 75% of students passed an exam, it means 75 out of every 100 students passed.

4. **Quartiles**:
   - Quartiles are also a specific type of quantile.
   - They divide a dataset into four equal parts.
   - Quartiles range from 0 to 4.
   - The 1st quartile (25th percentile) separates the bottom 25% of values from the top 75%.
   - The 2nd quartile (50th percentile) is the median, dividing the data into equal halves.
   - The 3rd quartile (75th percentile) separates the bottom 75% of values from the top 25%.

In summary:
- **Quantiles** are a general concept that includes percentiles and quartiles.
- **Percentiles** divide data into 100 equal parts (e.g., 90th percentile).
- **Quartiles** divide data into 4 equal parts (e.g., 1st quartile, 2nd quartile, 3rd quartile).

`

* * *

Percentiles
-----------

Percentiles are used in statistics to give you a number that describes the value that a given percent of the values are lower than.

If we take a look at the values of the `wt` (weight) variable from the `mtcars` data set:

#### Observation of wt (weight)

<table><tbody><tr><td>1.513</td><td>1.615</td><td>1.835</td><td>1.935</td><td>2.140</td><td>2.200</td><td>2.320</td><td>2.465</td></tr><tr><td>2.620</td><td>2.770</td><td>2.780</td><td>2.875</td><td>3.150</td><td>3.170</td><td>3.190</td><td>3.215</td></tr><tr><td>3.435</td><td>3.440</td><td>3.440</td><td>3.440</td><td>3.460</td><td>3.520</td><td>3.570</td><td>3.570</td></tr><tr><td>3.730</td><td>3.780</td><td>3.840</td><td>3.845</td><td>4.070</td><td>5.250</td><td>5.345</td><td>5.424</td></tr></tbody></table>

What is the 75. percentile of the weight of the cars? The answer is 3.61 or 3 610 lbs, meaning that 75% or the cars weight 3 610 lbs or less:

### Example


```r
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Calculate the 75th percentile (third quartile) of the 'wt' (weight) column
quantile(Data_Cars$wt, c(0.75))
```

Here's what each part of the code does:

1. `Data_Cars <- mtcars`: This line creates a new variable called `Data_Cars` and assigns the entire `mtcars` data set to it. The `mtcars` data set contains information about various car models, including attributes like miles per gallon (mpg), horsepower, and weight.

2. `quantile(Data_Cars$wt, c(0.75))`: This expression calculates the **75th percentile** (also known as the third quartile) of the "wt" (weight) column in the `Data_Cars` data frame. The `c(0.75)` argument specifies that we want to find the value at the 75th percentile.

   The result will be a single numeric value representing the weight value at the 75th percentile. For example, if the 75th percentile weight is 3.325 (units depend on the original data), the output will be:
   ```
   75%
   3.325
   ```

By running this code, you'll get the weight value at the 75th percentile in the `mtcars` data set. 🚗🔢


[Try it Yourself »](https://www.w3schools.com/r/tryr.asp?filename=demo_stat_percentile)



If you run the `quantile()` function without specifying the `c()` parameter, you will get the percentiles of 0, 25, 50, 75 and 100:

### Example

```r
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Calculate the quantiles of the 'wt' (weight) column
quantile(Data_Cars$wt)
```

Here's what each part of the code does:

1. `Data_Cars <- mtcars`: This line creates a new variable called `Data_Cars` and assigns the entire `mtcars` data set to it. The `mtcars` data set contains information about various car models, including attributes like miles per gallon (mpg), horsepower, and weight.

2. `quantile(Data_Cars$wt)`: This expression calculates the **quantiles** of the "wt" (weight) column in the `Data_Cars` data frame. Quantiles divide the data into equal groups based on specified probabilities.

   The result will be a numeric vector containing the quantile values. For example:
   ```
   0%  25%  50%  75% 100%
   1.513 2.581 3.325 3.610 5.424
   #1.513 min weight
   #5.424 max weight
   ```

   These values represent the minimum, first quartile (25th percentile), median (50th percentile), third quartile (75th percentile), and maximum weight values in the data set.

By running this code, you'll get the quantile values for the weight of the cars in the `mtcars` data set. 🚗🔢


[Try it Yourself »](https://www.w3schools.com/r/tryr.asp?filename=demo_stat_percentile2)

* * *



  


In [ ]:
# Create a variable named Data_Cars and assign the mtcars data set to it
Data_Cars <- mtcars

# Calculate the quantiles of the 'wt' (weight) column. Lay out may be different from the above.
quantile(Data_Cars$wt)


0%     25%     50%     75%    100% 
1.51300 2.58125 3.32500 3.61000 5.42400

Quartiles
---------

Quartiles are data divided into four parts, when sorted in an ascending order:

1.  The value of the first quartile cuts off the first 25% of the data
2.  The value of the second quartile cuts off the first 50% of the data
3.  The value of the third quartile cuts off the first 75% of the data
4.  The value of the fourth quartile cuts off the 100% of the data

Use the `quantile()` function to get the quartiles.

#Exercises for Quartiles

## R Exercises on Percentiles and Quartiles:

This response combines the strengths of both Response A and Response B, addresses their shortcomings based on the ratings, and provides clear, well-structured exercises with detailed explanations:

**Exercise 1: Fuel Efficiency Quartiles (Easy)**

**Instructions:**

1. **Load the `mtcars` dataset:**

```R
library(datasets)
data(mtcars)
```

2. **Calculate fuel efficiency (MPG) percentiles:**

```R
# Find the 25th, 50th (median), and 75th percentile MPG values
mpg_quartiles <- quantile(mtcars$mpg, c(0.25, 0.5, 0.75))

# Interpret the results
print("Quartiles of MPG:")
print(paste0("  - 25th percentile:", mpg_quartiles[1], "mpg"))
print(paste0("  - Median (50th percentile):", mpg_quartiles[2], "mpg"))
print(paste0("  - 75th percentile:", mpg_quartiles[3], "mpg"))
```

3. **Challenge:** Create a boxplot showing the MPG distribution, highlighting the quartiles with vertical lines. Label the axes and add a title.

**Explanation:**

- `quantile()` is used to calculate the percentiles, specifying specific probabilities (0.25, 0.5, 0.75) for the desired quartiles.
- The results are interpreted, explaining what each percentile value represents.

**Exercise 2: Horsepower Heroes (Medium)**

**Instructions:**

1. **Filter high-horsepower cars:**

```R
# Select cars with horsepower (hp) exceeding 200
high_hp_cars <- mtcars[mtcars$hp > 200, ]
```

2. **Compare horsepower quartiles:**

```R
# Calculate horsepower quartiles for all and high-hp cars
all_hp_quartiles <- quantile(mtcars$hp)
high_hp_quartiles <- quantile(high_hp_cars$hp)

# Compare and print results
print("Horsepower Quartiles:")
print("  - All cars:")
print(paste0("      - 25th percentile:", all_hp_quartiles[1], "hp"))
print(paste0("      - Median (50th percentile):", all_hp_quartiles[2], "hp"))
print(paste0("      - 75th percentile:", all_hp_quartiles[3], "hp"))
print("  - High-horsepower cars:")
print(paste0("      - 25th percentile:", high_hp_quartiles[1], "hp"))
print(paste0("      - Median (50th percentile):", high_hp_quartiles[2], "hp"))
print(paste0("      - 75th percentile:", high_hp_quartiles[3], "hp"))
```

3. **Challenge:** Create a double violin plot or side-by-side boxplot to visually compare the horsepower distributions of all cars and high-hp cars. Label the axes and add a title.

**Explanation:**

- We filter cars with high horsepower and calculate their quartiles separately.
- The results are compared, showcasing the spread of horsepower in each group.

**Exercise 3: Gear Up or Shift Down (Hard)**

**Instructions:**

1. **Analyze transmission types:**

```R
# Count cars with automatic (A) and manual (M) transmissions
transmission_counts <- table(mtcars$am)
```

2. **Explore quartiles for gears and MPG:**

```R
# Calculate gear and MPG quartiles for each transmission type
gear_quartiles <- tapply(mtcars$gear, mtcars$am, quantile)
mpg_quartiles <- tapply(mtcars$mpg, mtcars$am, quantile)

# Print results
print("Gear Quartiles:")
print(paste0("  - Automatic:", gear_quartiles(A)))
print(paste0("  - Manual:", gear_quartiles(M)))

print("MPG Quartiles:")
print(paste0("  - Automatic:", mpg_quartiles(A)))
print(paste0("  - Manual:", mpg_quartiles(M)))
```

3. **Challenge:** Create a scatter plot with transparency or color-coding to show the relationship between gears and MPG, considering transmission type. Add relevant labels and a title.

**

#Solutions to Quartiles

**Exercise 1: Fuel Efficiency Quartiles**

```r
# Load the mtcars dataset
library(datasets)
data(mtcars)

# Calculate MPG percentiles
mpg_quartiles <- quantile(mtcars$mpg, c(0.25, 0.5, 0.75))

# Interpret the results
print("Quartiles of MPG:")
print(paste0("  - 25th percentile:", mpg_quartiles[1], "mpg"))
print(paste0("  - Median (50th percentile):", mpg_quartiles[2], "mpg"))
print(paste0("  - 75th percentile:", mpg_quartiles[3], "mpg"))

# Challenge: Boxplot
ggplot(mtcars, aes(x = mpg)) +
  geom_boxplot(notch = TRUE) +
  labs(title = "MPG Distribution with Quartiles", x = "MPG", y = "") +
  scale_x_continuous(limits = c(min(mtcars$mpg), max(mtcars$mpg))) +
  geom_vline(xintercept = mpg_quartiles, linetype = "dashed", color = "red")
```

**Explanation:**

- The code calculates the desired percentiles using `quantile()`.
- The boxplot shows the overall distribution with quartiles highlighted.

**Exercise 2: Horsepower Heroes**

```r
# Filter high-horsepower cars
high_hp_cars <- mtcars[mtcars$hp > 200, ]

# Calculate horsepower quartiles
all_hp_quartiles <- quantile(mtcars$hp)
high_hp_quartiles <- quantile(high_hp_cars$hp)

# Compare and print results
print("Horsepower Quartiles:")
print("  - All cars:")
print(paste0("      - 25th percentile:", all_hp_quartiles[1], "hp"))
print(paste0("      - Median (50th percentile):", all_hp_quartiles[2], "hp"))
print(paste0("      - 75th percentile:", all_hp_quartiles[3], "hp"))
print("  - High-horsepower cars:")
print(paste0("      - 25th percentile:", high_hp_quartiles[1], "hp"))
print(paste0("      - Median (50th percentile):", high_hp_quartiles[2], "hp"))
print(paste0("      - 75th percentile:", high_hp_quartiles[3], "hp"))

# Challenge: Double violin plot
library(ggplot2)
ggplot(mtcars, aes(x = am, y = hp)) +
  geom_violin() +
  labs(title = "Horsepower Distribution by Transmission", x = "Transmission (A: Auto, M: Manual)", y = "Horsepower") +
  facet_wrap(~am)
```

**Explanation:**

- The code filters high-hp cars, calculates quartiles for both groups, and compares them.
- The violin plot visually shows the distributions and potential differences.

**Exercise 3: Gear Up or Shift Down**

```r
# Analyze transmission types
transmission_counts <- table(mtcars$am)

# Calculate gear and MPG quartiles for each transmission
gear_quartiles <- tapply(mtcars$gear, mtcars$am, quantile)
mpg_quartiles <- tapply(mtcars$mpg, mtcars$am, quantile)

# Print results
print("Gear Quartiles:")
print(paste0("  - Automatic:", gear_quartiles(A)))
print(paste0("  - Manual:", gear_quartiles(M)))

print("MPG Quartiles:")
print(paste0("  - Automatic:", mpg_quartiles(A)))
print(paste0("  - Manual:", mpg_quartiles(M)))

# Challenge: Scatter plot with color-coding
ggplot(mtcars, aes(x = gear, y = mpg, color = factor(am))) +
  geom_point(alpha = 0.5) +
  labs(title = "Gears vs. MPG by Transmission", x = "Number of Gears", y = "MPG") +
  scale_color_discrete(labels = c("Automatic", "Manual"))
```



**Exercise 3: Gear Up or Shift Down**

**Instructions:**

1. **Analyze transmission types:**

```r
# Analyze transmission types
transmission_counts <- table(mtcars$am)
```

2. **Explore quartiles for gears and MPG:**

```r
# Calculate gear and MPG quartiles for each transmission type
gear_quartiles <- tapply(mtcars$gear, mtcars$am, quantile)
mpg_quartiles <- tapply(mtcars$mpg, mtcars$am, quantile)

# Print results
print("Gear Quartiles:")
print(paste0("  - Automatic:", gear_quartiles(A)))
print(paste0("  - Manual:", gear_quartiles(M)))

print("MPG Quartiles:")
print(paste0("  - Automatic:", mpg_quartiles(A)))
print(paste0("  - Manual:", mpg_quartiles(M)))
```

3. **Challenge:** Create a scatter plot with transparency or color-coding to show the relationship between gears and MPG, considering transmission type. Add relevant labels and a title.

```r
# Challenge: Scatter plot with color-coding

# Load ggplot2 library
library(ggplot2)

# Create scatter plot with transparency and color-coding
ggplot(mtcars, aes(x = gear, y = mpg, color = factor(am))) +
  geom_point(alpha = 0.5) +  # Add transparency for better visibility
  labs(title = "Gears vs. MPG by Transmission", x = "Number of Gears", y = "MPG") +
  scale_color_discrete(labels = c("Automatic", "Manual"))

# Customize the plot (optional)
theme_bw()  # Black and white theme for clarity
```

**Explanation:**

1. The code calculates gear and MPG quartiles separately for automatic and manual transmissions.
2. The challenge creates a scatter plot with "factor(am)" defining color by transmission type.
3. Transparency (alpha) is added for better overlapping point visualization.
4. Labels and title are added for clarity.
5. An optional `theme_bw()` is used for a cleaner black and white theme.

Feel free to modify this code further to explore the data and customize the plot to your preferences.